File: dlitscustom/dlitscustom/report/pricing_rules_analysis/pricing_rules_analysis.py
# Copyright (c) 2024, DLITS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, today

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "label": _("Pricing Rule"),
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Pricing Rule Dlits",
            "width": 150
        },
        {
            "label": _("Enabled"),
            "fieldname": "enabled",
            "fieldtype": "Check",
            "width": 80
        },
        {
            "label": _("Priority"),
            "fieldname": "priority",
            "fieldtype": "Int",
            "width": 80
        },
        {
            "label": _("Price List"),
            "fieldname": "price_list",
            "fieldtype": "Link",
            "options": "Price List",
            "width": 120
        },
        {
            "label": _("Base Price Type"),
            "fieldname": "base_price_type",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Discount Type"),
            "fieldname": "discount_type",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Discount Value"),
            "fieldname": "discount_value",
            "fieldtype": "Float",
            "width": 100
        },
        {
            "label": _("Apply On"),
            "fieldname": "apply_on",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Apply To"),
            "fieldname": "apply_to",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Items Count"),
            "fieldname": "items_count",
            "fieldtype": "Int",
            "width": 100
        },
        {
            "label": _("Customers Count"),
            "fieldname": "customers_count",
            "fieldtype": "Int",
            "width": 120
        },
        {
            "label": _("Usage Count"),
            "fieldname": "usage_count",
            "fieldtype": "Int",
            "width": 100
        },
        {
            "label": _("Last Used"),
            "fieldname": "last_used",
            "fieldtype": "Date",
            "width": 100
        },
        {
            "label": _("Created On"),
            "fieldname": "creation",
            "fieldtype": "Datetime",
            "width": 120
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    
    # Get pricing rules with counts
    data = frappe.db.sql(f"""
        SELECT 
            pr.name,
            pr.enabled,
            pr.priority,
            pr.price_list,
            pr.base_price_type,
            pr.discount_type,
            pr.discount_value,
            pr.apply_on,
            pr.apply_to,
            pr.creation,
            
            -- Count items
            (SELECT COUNT(*) FROM `tabPricing Rule Item Code Dlits` 
             WHERE parent = pr.name) as items_count,
            
            -- Count customers
            (SELECT COUNT(*) FROM `tabPricing Rule Customer Dlits` 
             WHERE parent = pr.name) as customers_count,
            
            -- Count item groups
            (SELECT COUNT(*) FROM `tabPricing Rule Item Group Dlits` 
             WHERE parent = pr.name) as item_groups_count,
            
            -- Count brands
            (SELECT COUNT(*) FROM `tabPricing Rule Brand Dlits` 
             WHERE parent = pr.name) as brands_count,
            
            -- Count customer groups
            (SELECT COUNT(*) FROM `tabPricing Rule Customer Group Dlits` 
             WHERE parent = pr.name) as customer_groups_count
             
        FROM `tabPricing Rule Dlits` pr
        WHERE 1=1 {conditions}
        ORDER BY pr.priority ASC, pr.creation DESC
    """, as_dict=True)
    
    # Get usage statistics for each pricing rule
    for row in data:
        usage_stats = get_pricing_rule_usage_stats(row.name, filters)
        row.update(usage_stats)
        
        # Calculate total apply targets
        if row.apply_on == "Item Code":
            row.apply_targets_count = row.items_count
        elif row.apply_on == "Item Group":
            row.apply_targets_count = row.item_groups_count
        elif row.apply_on == "Brand":
            row.apply_targets_count = row.brands_count
        
        if row.apply_to == "Customer":
            row.apply_to_count = row.customers_count
        elif row.apply_to == "Customer Group":
            row.apply_to_count = row.customer_groups_count
    
    return data

def get_conditions(filters):
    conditions = ""
    
    if filters.get("enabled"):
        conditions += " AND pr.enabled = 1"
    
    if filters.get("price_list"):
        conditions += f" AND pr.price_list = '{filters.get('price_list')}'"
    
    if filters.get("base_price_type"):
        conditions += f" AND pr.base_price_type = '{filters.get('base_price_type')}'"
    
    if filters.get("apply_on"):
        conditions += f" AND pr.apply_on = '{filters.get('apply_on')}'"
    
    if filters.get("apply_to"):
        conditions += f" AND pr.apply_to = '{filters.get('apply_to')}'"
    
    if filters.get("from_date"):
        conditions += f" AND DATE(pr.creation) >= '{filters.get('from_date')}'"
    
    if filters.get("to_date"):
        conditions += f" AND DATE(pr.creation) <= '{filters.get('to_date')}'"
    
    return conditions

def get_pricing_rule_usage_stats(pricing_rule_name, filters):
    """Get usage statistics for a pricing rule"""
    
    # Get date range for usage stats
    from_date = filters.get("usage_from_date") or frappe.utils.add_months(today(), -3)
    to_date = filters.get("usage_to_date") or today()
    
    # This is a simplified approach - in a real implementation, you would need to
    # track pricing rule usage in sales transactions
    
    # For now, return placeholder data
    return {
        "usage_count": 0,
        "last_used": None
    }

def get_chart_data(data, filters):
    """Generate chart data for the report"""
    
    # Base Price Type Distribution
    base_price_types = {}
    enabled_count = 0
    disabled_count = 0
    
    for row in data:
        # Count by base price type
        base_price_type = row.get("base_price_type", "Unknown")
        base_price_types[base_price_type] = base_price_types.get(base_price_type, 0) + 1
        
        # Count enabled/disabled
        if row.get("enabled"):
            enabled_count += 1
        else:
            disabled_count += 1
    
    return {
        "data": {
            "labels": list(base_price_types.keys()),
            "datasets": [
                {
                    "name": "Pricing Rules by Base Price Type",
                    "values": list(base_price_types.values())
                }
            ]
        },
        "type": "donut",
        "height": 300
    }

File: dlitscustom/dlitscustom/report/pricing_rules_analysis/pricing_rules_analysis.js
// Copyright (c) 2024, DLITS and contributors
// For license information, please see license.txt

frappe.query_reports["Pricing Rules Analysis"] = {
	"filters": [
		{
			"fieldname": "enabled",
			"label": __("Show Only Enabled"),
			"fieldtype": "Check",
			"default": 0
		},
		{
			"fieldname": "price_list",
			"label": __("Price List"),
			"fieldtype": "Link",
			"options": "Price List"
		},
		{
			"fieldname": "base_price_type",
			"label": __("Base Price Type"),
			"fieldtype": "Select",
			"options": "\nPrice List Rate\nValuation Rate\nSelling Rate\nItem Rate"
		},
		{
			"fieldname": "apply_on",
			"label": __("Apply On"),
			"fieldtype": "Select",
			"options": "\nItem Code\nItem Group\nBrand"
		},
		{
			"fieldname": "apply_to",
			"label": __("Apply To"),
			"fieldtype": "Select",
			"options": "\nCustomer\nCustomer Group"
		},
		{
			"fieldname": "from_date",
			"label": __("Created From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -3)
		},
		{
			"fieldname": "to_date",
			"label": __("Created To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname": "usage_from_date",
			"label": __("Usage From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1)
		},
		{
			"fieldname": "usage_to_date",
			"label": __("Usage To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today()
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (column.fieldname == "enabled" && data && data.enabled == 1) {
			value = `<span class="indicator green">Enabled</span>`;
		} else if (column.fieldname == "enabled" && data && data.enabled == 0) {
			value = `<span class="indicator red">Disabled</span>`;
		}
		
		if (column.fieldname == "base_price_type" && data) {
			let color = "blue";
			if (data.base_price_type == "Item Rate") color = "purple";
			else if (data.base_price_type == "Selling Rate") color = "green";
			else if (data.base_price_type == "Valuation Rate") color = "orange";
			
			value = `<span class="indicator ${color}">${data.base_price_type}</span>`;
		}
		
		if (column.fieldname == "priority" && data && data.priority == 0) {
			value = `<span class="text-muted">Default</span>`;
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Add custom buttons
		report.page.add_inner_button(__("Create New Pricing Rule"), function() {
			frappe.new_doc("Pricing Rule Dlits");
		});
		
		report.page.add_inner_button(__("Pricing Rule Settings"), function() {
			frappe.set_route("List", "Pricing Rule Dlits");
		});
		
		report.page.add_inner_button(__("Test Pricing Rule"), function() {
			let d = new frappe.ui.Dialog({
				title: __("Test Pricing Rule"),
				fields: [
					{
						fieldname: "item_code",
						label: __("Item Code"),
						fieldtype: "Link",
						options: "Item",
						reqd: 1
					},
					{
						fieldname: "customer",
						label: __("Customer"),
						fieldtype: "Link",
						options: "Customer"
					},
					{
						fieldname: "price_list",
						label: __("Price List"),
						fieldtype: "Link",
						options: "Price List",
						default: "Standard Selling"
					}
				],
				primary_action_label: __("Test"),
				primary_action: function(values) {
					frappe.call({
						method: "dlitscustom.utils.get_pricing_rule_dlits.get_pricing_rule_dlits",
						args: {
							item_code: values.item_code,
							customer: values.customer,
							price_list: values.price_list
						},
						callback: function(r) {
							if (r.message) {
								frappe.msgprint({
									title: __("Pricing Rule Result"),
									message: __("Calculated Price: {0}", [r.message]),
									indicator: "green"
								});
							} else {
								frappe.msgprint({
									title: __("No Pricing Rule"),
									message: __("No pricing rule found for the given criteria"),
									indicator: "orange"
								});
							}
						}
					});
					d.hide();
				}
			});
			d.show();
		});
	}
};